Add `display_name` to UserProfile

emdiso opened the pull request:
made requested changes from: https://github.com/e-valuation/EvaP/pull/1895

- removed requested files
- removed endpoint making a manual database entry
- removed edit display name modal
- changed the profile page table to a form, except for the CC users part which is simply shown below the form 
- disabled all form fields except for display name
- added display_name field to staff user form as well 
- added logic to profile_edit function in /evaluation/views.py to save changes

let me know if it all looks ok. very new to Django :)

janno42 commented on evap/contributor/tests/test_views.py:
You can revert the changes in this file because the button doesn't exist anymore.

janno42 commented on evap/evaluation/templates/profile.html:
I wanted to suggest multiple changes to the UI and there was a bug with the two forms (when changing delegates, the display name would be deleted). This turned out to be quite a lot and while figuring out the UI, I had to build it anyway. So I put all the changes into a commit and pushed it to your branch. You're welcome to use it as-is.

janno42 reviewed (changes requested):
Good work on the form! Again, I'm sorry this is so much work in your first PR but I'm glad you found a good solution!
There are a few more thing to address though. Please let us know if you need any further information.

emdiso commented:
Hey, I think I addressed everything except for this test:
In evap/evaluation/tests/test_views.py <https://github.com/e-valuation/EvaP/pull/1903#discussion_r1142470178>:

> -        self.assertIn(user.email, page)
+        # self.assertIn(user.email, page) # wont be in html anymore
The email is still shown on the page (as it should be), so this needs to be reverted.

It seems like the test is expecting the email to appear in the HTML template itself as text, which it doesn’t anymore since it is shown on the page dynamically. I am not too sure how to address this. 

Let me know if there is anything else/ if I hit everything necessary. Thanks!


> On Mar 20, 2023, at 3:40 PM, Johannes Wolf ***@***.***> wrote:
> 
> 
> @janno42 requested changes on this pull request.
> 
> Good work on the form! Again, I'm sorry this is so much work in your first PR but I'm glad you found a good solution!
> There are a few more thing to address though. Please let us know if you need any further information.
> 
> In evap/evaluation/forms.py <https://github.com/e-valuation/EvaP/pull/1903#discussion_r1142428776>:
> 
> > +        self.fields["title"].initial = self.instance.title
> +        self.fields["display_name"].initial = self.instance.display_name
> +        self.fields["first_name"].initial = self.instance.first_name
> +        self.fields["last_name"].initial = not self.instance.last_name
> +        self.fields["email"].initial = not self.instance.email
> For each field except display_name please add self.fields["title"].disabled = "True" and similar
> 
> In evap/evaluation/forms.py <https://github.com/e-valuation/EvaP/pull/1903#discussion_r1142428916>:
> 
> > +    title = forms.CharField(max_length=30, required=False, label=_("Title"), disabled=True)
> +    display_name = forms.CharField(max_length=30, required=False, label=_("Display name"))
> +    first_name = forms.CharField(max_length=30, required=False, label=_("First name"), disabled=True)
> +    last_name = forms.CharField(max_length=30, required=False, label=_("Last name"), disabled=True)
> +    email = forms.CharField(max_length=30, required=False, label=_("Email"), disabled=True)
> Django makes it easy to define forms by automatically creating fields based on the model's specification when using ModelForms like in this case.
> Here you are creating a ModelForm for a UserProfile. By defining which fields to use below (listing the fields in the order in which they should be shown in the form) Django knows that, e.g., the title should be a CharField. It is important to note that in the UserProfile model definition in evaluation/models.py we define max_length=255 for title and by letting Django handle the form creation, you don't have to redefine it and (unintentionally) redefine the length to 30 characters.
> 
> So here you can remove all these lines. The only missing thing then is the disabled definition. This can be done at __init__ below, I added a comment there.
> 
> In evap/evaluation/forms.py <https://github.com/e-valuation/EvaP/pull/1903#discussion_r1142468412>:
> 
> > +        self.fields["title"].initial = self.instance.title
> +        self.fields["display_name"].initial = self.instance.display_name
> +        self.fields["first_name"].initial = self.instance.first_name
> +        self.fields["last_name"].initial = not self.instance.last_name
> +        self.fields["email"].initial = not self.instance.email
> ⬇️ Suggested change
> -        self.fields["title"].initial = self.instance.title
> -        self.fields["display_name"].initial = self.instance.display_name
> -        self.fields["first_name"].initial = self.instance.first_name
> -        self.fields["last_name"].initial = not self.instance.last_name
> -        self.fields["email"].initial = not self.instance.email
> These are not needed, Django will fill all fields with the object's values.
> 
> In evap/evaluation/tests/test_views.py <https://github.com/e-valuation/EvaP/pull/1903#discussion_r1142470178>:
> 
> > -        self.assertIn(user.email, page)
> +        # self.assertIn(user.email, page) # wont be in html anymore
> The email is still shown on the page (as it should be), so this needs to be reverted.
> 
> On package-lock.json <https://github.com/e-valuation/EvaP/pull/1903#discussion_r1142478089>:
> 
> This was a misunderstanding: Please do not delete the file and commit the deletion but instead remove this file from your commits.
> (Or to make things easier for you: Re-add the file as we will probably squash all your changes into one commit when merging.)
> 
> On package.json <https://github.com/e-valuation/EvaP/pull/1903#discussion_r1142478452>:
> 
> Same here
> 
> On evap/contributor/tests/test_views.py <https://github.com/e-valuation/EvaP/pull/1903#discussion_r1142583578>:
> 
> You can revert the changes in this file because the button doesn't exist anymore.
> 
> On evap/evaluation/templates/profile.html <https://github.com/e-valuation/EvaP/pull/1903#discussion_r1142588255>:
> 
> I wanted to suggest multiple changes to the UI and there was a bug with the two forms (when changing delegates, the display name would be deleted). This turned out to be quite a lot and while figuring out the UI, I had to build it anyway. So I put all the changes into a commit and pushed it to your branch. You're welcome to use it as-is.
> 
> In evap/staff/forms.py <https://github.com/e-valuation/EvaP/pull/1903#discussion_r1142593142>:
> 
> > @@ -1020,7 +1020,7 @@ def save(self, *args, **kw):
>          )
>  
>          # refresh results cache
> This is now also relevant for the profile page. Before, users couldn't change any cached data, but now we have to recache the results in case the display name of a user changed.
> Can you add this in the evaluation/forms.py as well?
> 
> —
> Reply to this email directly, view it on GitHub <https://github.com/e-valuation/EvaP/pull/1903#pullrequestreview-1348984716>, or unsubscribe <https://github.com/notifications/unsubscribe-auth/AYPCN4SD6L23TUEARZUAOUTW5CXCJANCNFSM6AAAAAAV4NGPCE>.
> You are receiving this because you authored the thread.
>

janno42 reviewed (commented):
The test works fine for me. Actually, it sends a GET request (`self.app.get`) and then checks for text within the rendered page. Thus, the email field is filled and the user's email can be found on the page. So you can remove the comment and everything should be working :)

janno42 commented on evap/evaluation/forms.py:
```diff
@@ -118,3 +119,30 @@ class Meta:
+                contributions__contributor=self.instance, state__in=STATES_WITH_RESULTS_CACHING
+            ).distinct()
+            for evaluation in evaluations:
+                cache_results(evaluation)
```
You are missing an import (see the linter's output)

janno42 commented:
I didn't do a complete final review yet, but there should only be minor things, if any.

janno42 reviewed (commented):
✔️ Meets requirements
✔️ UI functionality checked

He3lixxx reviewed (changes requested):
Hey, thanks for working on this!

In an offline discussion, @janno42 and me found a few more spots that need to be changed that we forgot to list in the issue. I've added annotations for these starting with "(follow-up)". In order to not bloat the scope of this PR, we'd say its fine to ignore these for now, merge this pull request, and then create a follow up issue for these. However, if you'd already like to fix some of them here, you're also very welcome to do so.

In general, we much appreciate the work so far, and in case you don't have time to keep working on this, we're also happy to take over and implement the remaining changes ourselves. Just give us a small ping in this case, so we know we can go ahead.

niklasmohrin commented on evap/evaluation/views.py:
```diff
@@ -198,29 +198,34 @@ def set_lang(request):
+    if request.method == "POST":
+        operation = request.POST.get("operation")
+        if operation not in ("profile", "delegates"):
+            raise SuspiciousOperation("Invalid POST operation")
```
Actually, I think it would be better to create two new views for the two forms on the profile site, maybe `/profile/edit` and `/profile/request_changes`. This way, we get rid of the `operation` parameter :)

He3lixxx replied:
@niklasmohrin I did not make this into two views for now, as discussed offline, it's a bit annoying to do. Resolving this discussion for now. Feel free to open another issue for that (would then also affect staff import persons view).

janno42 reviewed (approved):
✔️ Meets requirements
✔️ UI functionality checked

niklasmohrin commented on evap/results/tools.py:
```diff
@@ -451,7 +451,7 @@ def textanswers_visible_to(contribution):
     num_delegates = len(delegates - contributors)
 
-    sorted_contributors = sorted(contributors, key=lambda user: (user.last_name, user.first_name))
+    sorted_contributors = sorted(contributors, key=lambda user: (user.ordering_key))
```
Maybe we shouldn't make it a property so that we can use `UserProfile.ordering_key` instead of a lambda here?

niklasmohrin commented on evap/evaluation/management/commands/anonymize.py:
```diff
@@ -115,6 +115,7 @@ def anonymize_users(self, first_names, last_names):
             if user.email and user.email.split("@")[0] in Command.ignore_email_usernames:
                 continue
             user.first_name = name[0]
+            user.display_name = random.choice(first_names) if random.randint(0, 100) < 10 else ""  # nosec
```
This seems odd, let's go for `random.random() < 0.1`?

niklasmohrin commented on evap/evaluation/migrations/0133_userprofile_display_name.py:
```diff
@@ -0,0 +1,34 @@
+
+class Migration(migrations.Migration):
+    dependencies = [
+        ("evaluation", "0132_textanswer_is_flagged"),
```
Infotexts took place 133 :^)

niklasmohrin reviewed (commented):
some comments (to be continued)

He3lixxx commented on evap/staff/views.py:
```diff
@@ -2071,7 +2071,7 @@ def user_list(request):
             "courses_responsible_for",
         )
-        .order_by("last_name", "first_name", "email")
+        .order_by(*UserProfile._meta.ordering)
```
for some weird reason, it isn't -- but I haven't invested the time to find out why. Probably due to the complex query building above?

niklasmohrin commented on evap/evaluation/models.py:
```diff
@@ -1639,7 +1646,13 @@ class UserProfile(AbstractBaseUser, PermissionsMixin):
 
     class Meta:
-        ordering = [Lower("last_name"), Lower("first_name"), Lower("email")]
+        # keep in sync with ordering_key property
```
```suggestion
        # keep in sync with ordering_key method
```

niklasmohrin commented on evap/evaluation/models.py:
```diff
@@ -1654,11 +1667,20 @@ def save(self, *args, **kwargs):
+    def ordering_key(self):
+        # keep in sync with Meta.ordering
+        lower_last_name = (self.last_name or "").lower()
+        lower_display_or_first_name = ((self.display_name if self.display_name else self.first_name) or "").lower()
```
How do you feel about this?
```suggestion
        lower_display_or_first_name = (self.display_name or self.first_name or "").lower()
```

niklasmohrin commented on evap/evaluation/tests/test_views.py:
```diff
@@ -156,17 +156,34 @@ def test_requires_login(self):
+        form = page.forms["profile-form"]
+        form["display_name"] = "testdisplayname"
+        form.submit(name="operation", value="profile")
+        self.assertTrue(UserProfile.objects.filter(display_name="testdisplayname").exists())
```
should we make sure it doesn't exist beforehand? the name seems like someone else could come up with it independently

niklasmohrin commented on evap/evaluation/views.py:
```diff
@@ -198,29 +198,34 @@ def set_lang(request):
-            },
-        )
+    profile_form = ProfileForm(request.POST or None, request.FILES or None, instance=user)
+    form = DelegatesForm(request.POST or None, request.FILES or None, instance=user)
```
Maybe rename it to `delegates_form` now that there are two?

niklasmohrin commented on evap/evaluation/views.py:
```diff
@@ -198,29 +198,34 @@ def set_lang(request):
         request,
         "profile.html",
-        {"user": user},
+        {
+            "user": user,
+            "form": form,
+            "profile_form": profile_form,
+            "delegate_of": user.represented_users.all(),
+            "cc_users": user.cc_users.all(),
+            "ccing_users": user.ccing_users.all(),
+        },
```
These now also get passed if the user is not an editor - it probably doesn't matter, but let's make this tight and leave them out for normal users?

niklasmohrin commented on evap/evaluation/templates/profile.html:
```diff
@@ -9,50 +9,45 @@ <h3>{% trans 'Profile' %}</h3>
+                <form id="profile-form" method="POST" enctype="multipart/form-data" class="form-horizontal multiselect-form">
+                    <div class="card-body">
+                        <h4 class="card-title mb-4">{% trans 'Personal information' %}</h4>
+                        {% csrf_token %}
```
Nit: Let's move the csrf token to right after the form :^)

niklasmohrin reviewed (commented):
Looks mostly good, some nits. Haven't thought about whether there are more places yet

niklasmohrin reviewed (approved):
Looks good
